File: Rect.swift
//
//  Rect.swift
//  TestEngine
//
//  Created by Vladas Zakrevskis on 7/8/17.
//  Copyright © 2017 VladasZ. All rights reserved.
//

import Foundation

class Rect {
    
    var x: Double { get { return origin.x } set { origin.x = newValue } }
    var y: Double { get { return origin.y } set { origin.y = newValue } }
    var width:  Double { get { return size.width }  set { size.width = newValue } }
    var height: Double { get { return size.height } set { size.height = newValue } }
    
    var origin: Point = Point()
    var size:   Size  = Size()
    
    init() { }
    
    init(size: Size) {
        
        self.size = size
    }
    
    init(origin: Point, size: Size) {
        
        self.origin = origin
        self.size   = size
    }
    
    init(_ x: Double, _ y: Double, _ width: Double, _ height: Double) {
        
        origin = Point(x, y)
        size   = Size(width, height)
    }
}

File: source/Geometry/source/Primitives/Rect.cpp
//
//  Rect.cpp
//  TestEngine
//
//  Created by Vladas Zakrevskis on 8/28/17.
//  Copyright © 2017 VladasZ. All rights reserved.
//

#include "Rect.hpp"
#include "System.hpp"

Rect::Rect(float width, float height) : size(Size(width, height)) { }

Rect::Rect(const Size &_size) : size(_size) { }

Rect::Rect(float x, float y, float width, float height) : origin(Point(x, y)), size(Size(width, height)) { }

Rect& Rect::operator=(const Rect &r2) {
    origin = r2.origin;
    size = r2.size;
    return* this;
}

float Rect::maxX() const { return origin.x + size.width; }

float Rect::maxY() const { return origin.y + size.height; }

bool Rect::contains(const Point &point) const {
    return  point.x > origin.x &&
            point.y > origin.y &&
           (point.x < origin.x + size.width) &&
           (point.y < origin.y + size.height);
}

Rect Rect::withZeroOrigin() const {
	return { size.width, size.height };
}

String Rect::toString() const {
    return "x: "_s +  origin.x  +      " y: " + origin.y +
      " width: "   + size.width + " height: " + size.width;
}

Rect Rect::random() {
    static const int size = 300;
    return {
        (float)System::random(size),
        (float)System::random(size),
        (float)System::random(size),
        (float)System::random(size)
    };
}
